pkg/api/message/v1/service.go

```diff
@@ -35,6 +36,24 @@ const (
 
 	// 1048576 - 300 - 62 = 1048214
 	MaxMessageSize = pubsub.DefaultMaxMessageSize - MaxContentTopicNameSize - 62
+
+	// maxRequestLimitPerBatch defines the maximum number of request we can support per batch.
+	maxRequestLimitPerBatch = 50
+
+	// maxTopicsPerRequest defines the maximum number of topics that can be queried in a single request.
+	maxTopicsPerRequest = 1024
+
+	// maxTopicsPerBatch defines the maximum number of topics that can be queried in a batch query. This
+	// limit is imposed in additional to the per-query limit maxTopicsPerRequest.
+	maxTopicsPerBatch = 4096
```

Comment:
We have a [metric](https://app.datadoghq.com/dashboard/hmw-rv4-hxz/incident-15-investigation?fullscreen_end_ts=1707943281015&fullscreen_paused=false&fullscreen_refresh_mode=sliding&fullscreen_section=overview&fullscreen_start_ts=1707338481015&fullscreen_widget=8316307017714266&refresh_mode=sliding&view=spans&from_ts=1707856851255&to_ts=1707943251255&live=true) that shows up to 90k in the same request, so dropping it down to 4k would break that usage, and I doubt those clients have anything in place currently to auto-adjust down to something lower because it'd be a new error.

Curious what @neekolas thinks about having a limit here too.

Reply:
could you tell me more about these clients ? are these the notification servers ?

Reply:
The notification servers tend to use the SubscribeAll endpoint which just listens to the whole firehose rather than specific topics. These clients are more likely chatbots that listen to a lot of topics for their users. We could push them to use SubscribeAll but today they're listening to individual topics via this endpoint. I had a [wip branch](https://github.com/xmtp/xmtp-node-go/tree/snor/sub-nats-wildcard-for-many-topics) from a few weeks ago that updates this subscribe method to use the wildcard (listen to all topics) inside of this Subscribe endpoint to optimize memory usage (reduce open goroutines where there's a goroutine per topic subscription now), that could be worth implementing here instead of a limit on the batch size.

Reply:
ok, so I think that now I understand where the memory issue is coming from.

I was thinking about #topic ~= number of concurrent conversation in the inbox. but I wan't aware of chatbots.

These are completely different use case, with (potentially) different attributes : we might want to favor slower operation @ a lower resource consumption for that.

I'll spend some more time looking into that, but I do think your direction is the way to go.

Reply:
Agree with @snormore that a hard limit isn't an option. If we were to change the behaviour for existing apps, we'd want to give them a very generous warning and a reasonable alternative.

Reply:
would you be ok if we'll start off with a very high number, and add configurable options later on ? i.e. start with a 1M limit ( far beyond what would crash the server )

Reply:
The above metric mention by @snormore applies to the subscribe, not for the query, btw. We don't have a metric, afaik, that tracks the query size.

Reply:
I updated the numerics here per a unit-test created, where I tested the max possible topics count.
This would guarantee it's won't break anything for the time being, and we will be able to adjust it in the future.

Reply:
Something like 1M per request would be fine. It's way above what any valid user needs.

If we were to enforce the limit per IP, we'd just have to be very sure the bookkeeping is correct. For example, that a client can't subscribe to 100,000 topics, disconnect, and resubscribe to the same list and push themselves over the limit.

Reply:
Agree - IP filtering isn't trivial. But it's also a bit beyond the scope of this PR. The goal here was to apply some basic and immediate remedies to the existing (crashing) node on the "happy" path.
I can look into IP filtering next.